modules/library/render/src/main/java/org/geotools/renderer/label/LabelCacheImpl.java

```diff
@@ -1265,6 +1272,43 @@ private boolean paintPointLabelInternal(LabelPainter painter, AffineTransform te
             return true;
         }
     }
+    
+    /**
+     * Splits the polygon into its component lines and labels those, one by one
+     */
+    private boolean paintPolygonBorder(LabelPainter painter, AffineTransform tempTransform,
+            Rectangle displayArea, LabelIndex glyphs) throws Exception {
+        // turn the polygon in its component lines
+        Geometry geometry = painter.getLabel().getGeometry();
+        List<LineString> lines = new ArrayList<>();
+        geometry.apply(new GeometryComponentFilter() {
+
+            @Override
+            public void filter(Geometry g) {
+                if (g instanceof LineString) {
+                    lines.add((LineString) g);
+                }
+            }
+        });
```

Comment:
Heh, that was my first idea as well, I've tried for some time, but could not make it compile. From the looks of it, I was missing the cast to GeometryComponentyFilter.